### PR TITLE
Add @Unremoveable annotation to Arc

### PR DESCRIPTION
**Pull request by geoand**
This is one part of the solutions discussed in #2576

### CODE REVIEW

**Comment by geoand** [comment]
I will follow up later with a second PR for the other part of the solution we discussed in #2576

**Comment by stuartwdouglas** [comment]
LGTM, but @mkouba can merge it if he is happy with it.

File: independent-projects/arc/runtime/src/main/java/io/quarkus/arc/Unremovable.java [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+/**
+ * Indicates that the class marked with this annotation should never be removed by Arc as a bean
+ * even if it has no known injection points
+ */
+@Retention(RetentionPolicy.RUNTIME)
+@Target(ElementType.TYPE)
```
Comment (mkouba) [inline independent-projects/arc/runtime/src/main/java/io/quarkus/arc/Unremovable.java]: We should add `ElementType.METHOD` and `ElementType.FIELD`  and make it possible to annotate a producer method/field.

File: docs/src/main/asciidoc/cdi-reference.adoc [hunk 2]
```diff
@@ -283,5 +283,7 @@ An unused bean:
 * does not declare an observer,
 * does not declare any producer which is eligible for injection to any injection point,
 * is not directly eligible for injection into any `javax.enterprise.inject.Instance` or `javax.inject.Provider` injection point
+
+Users can force Arc to not remove any of their specific beans (even if they satisfy all the rules specified above) by annotating them with `io.quarkus.arc.Unremovable`.
```
Comment (mkouba) [inline docs/src/main/asciidoc/cdi-reference.adoc]: "force Arc" -> "instruct the container"

**Review by mkouba — CHANGES_REQUESTED** [review]

File: independent-projects/arc/runtime/src/main/java/io/quarkus/arc/Unremovable.java (hunk 1 above)
Comment (geoand) [reply independent-projects/arc/runtime/src/main/java/io/quarkus/arc/Unremovable.java]: Hadn't thought of that :). So in this case we would need to do some extra work to figure out which class is unremovable, correct?

Comment (geoand) [reply independent-projects/arc/runtime/src/main/java/io/quarkus/arc/Unremovable.java]: Should we do the extra processing work in this PR do you think or leave it as an enhancement for later?

File: docs/src/main/asciidoc/cdi-reference.adoc (hunk 2 above)
Comment (geoand) [reply docs/src/main/asciidoc/cdi-reference.adoc]: Fixed

File: independent-projects/arc/runtime/src/main/java/io/quarkus/arc/Unremovable.java (hunk 1 above)
Comment (geoand) [reply independent-projects/arc/runtime/src/main/java/io/quarkus/arc/Unremovable.java]: If we want it part of this PR, I can most likely spare some time to do it later on today.

Comment (mkouba) [reply independent-projects/arc/runtime/src/main/java/io/quarkus/arc/Unremovable.java]: Hm, it shouldn't require any extra work. Just discover all `@Unremovable` annotations and match the `AnnotationInstance.target()` with `BeanInfo.getTarget()` .

Comment (geoand) [reply independent-projects/arc/runtime/src/main/java/io/quarkus/arc/Unremovable.java]: Ah yes! I forgot that `AnnotationInstance.target()` would work for all types. I'll fix the PR soon then :)

Comment (geoand) [reply independent-projects/arc/runtime/src/main/java/io/quarkus/arc/Unremovable.java]: Fixed

File: extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/AnnotatedUnremovableClassesProcessor.java [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+import io.quarkus.deployment.annotations.BuildStep;
+
+public class AnnotatedUnremovableClassesProcessor {
+
+    private static final DotName UNREMOVABLE = DotName.createSimple(Unremovable.class.getName());
+
+    @BuildStep
+    public void findAnnotatedClasses(BeanArchiveIndexBuildItem beanArchiveIndexBuildItem,
+            BuildProducer<UnremovableBeanBuildItem> producer) {
+        final Set<AnnotationTarget> namesOfClassesMarkedAsUnremovable = new HashSet<>();
```
Comment (mkouba) [inline extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/AnnotatedUnremovableClassesProcessor.java]: Well, it's not `namesOfClassesMarkedAsUnremovable` anymore. Maybe `unremovableBeanTargets` or something like that?

File: extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/AnnotatedUnremovableClassesProcessor.java [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+        final Collection<AnnotationInstance> classesMarkedAsUnremovable = beanArchiveIndexBuildItem.getIndex()
```
Comment (mkouba) [inline extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/AnnotatedUnremovableClassesProcessor.java]: `classesMarkedAsUnremovable` is actually `unremovableAnnotations`

File: extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/AnnotatedUnremovableClassesProcessor.java (hunk 3 above)
Comment (geoand) [reply extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/AnnotatedUnremovableClassesProcessor.java]: Yes of course! I had set my mind to change it and then I forgot.... Sorry about this dump omission

File: extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/AnnotatedUnremovableClassesProcessor.java (hunk 4 above)
Comment (geoand) [reply extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/AnnotatedUnremovableClassesProcessor.java]: :+1:

File: independent-projects/arc/runtime/src/main/java/io/quarkus/arc/Unremovable.java [hunk 5]
```diff
@@ -0,0 +1,15 @@
+package io.quarkus.arc;
+
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+/**
+ * Indicates that the class marked with this annotation should never be removed by Arc as a bean
```
Comment (mkouba) [inline independent-projects/arc/runtime/src/main/java/io/quarkus/arc/Unremovable.java]: "Indicates that a bean marked with this annotation should never be removed."?

File: extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/AnnotatedUnremovableClassesProcessor.java (hunk 3 above)
Comment (geoand) [reply extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/AnnotatedUnremovableClassesProcessor.java]: Fixed

File: extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/AnnotatedUnremovableClassesProcessor.java (hunk 4 above)
Comment (geoand) [reply extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/AnnotatedUnremovableClassesProcessor.java]: Fixed

**Review by mkouba — APPROVED** [review]

File: independent-projects/arc/runtime/src/main/java/io/quarkus/arc/Unremovable.java (hunk 5 above)
Comment (geoand) [reply independent-projects/arc/runtime/src/main/java/io/quarkus/arc/Unremovable.java]: Fixed

**Review by mkouba — APPROVED** [review]

**Pull request merged**